Return source for fatal errors

Currently, when a parsing (`fatal`) error occurs, the `source` is not included in the message that ESLint returns.  It may be worthwhile to investigate whether the source could be returned, which is used in some formatters to show clearly where the problem is.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25500931-return-source-for-fatal-errors?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

Currently I see this when I run eslint on a invalid javascript code.

```sh
C:\Users\gs025879\Documents\webstrom\test\gyandeep.js
  3:18  error  Parsing error: Unexpected token )

✖ 1 problem (1 error, 0 warnings)
```

Is this what is expected or are we looking for something more than this?

--- a/docs/developer-guide/nodejs-api.md
+++ b/docs/developer-guide/nodejs-api.md
@@ -16,6 +16,27 @@ var code = new SourceCode("var foo = bar;", ast);
 
 The `SourceCode` constructor throws an error if the AST is missing any of the required information.
 
+### splitLines()
+
+This is a static function on `SourceCode` that is used to split the source code text into an array of lines.
+
+```js
+var SourceCode = require("eslint").SourceCode;
+
+var code = "var a = 1;\nvar b = 2;"
+
+// split code into an array
+var codeLines = SourceCode.splitLines(code);
+
+/*
+    Value of codeLines will be
+    [
+        "var a = 1;",
+        "var b = 2;"
+    ]
+ */
+```
+
 ## linter
 
 The `linter` object does the actual evaluation of the JavaScript code. It doesn't do any filesystem operations, it simply parses and reports on the code. You can retrieve `linter` like this:

--- a/lib/eslint.js
+++ b/lib/eslint.js
@@ -85,6 +85,7 @@ function parseJsonConfig(string, location, messages) {
             ruleId: null,
             fatal: true,
             severity: 2,
+            source: null,
             message: "Failed to parse JSON from '" + string + "': " + ex.message,
             line: location.start.line,
             column: location.start.column + 1
@@ -530,6 +531,7 @@ module.exports = (function() {
                 ruleId: null,
                 fatal: true,
                 severity: 2,
+                source: null,
                 message: ex.message,
                 line: 0,
                 column: 0
@@ -555,12 +557,13 @@ module.exports = (function() {
 
             // If the message includes a leading line number, strip it:
             var message = ex.message.replace(/^line \d+:/i, "").trim();
+            var source = (ex.lineNumber) ? SourceCode.splitLines(text)[ex.lineNumber - 1] : null;
 
             messages.push({
                 ruleId: null,
                 fatal: true,
                 severity: 2,
-
+                source: source,
                 message: "Parsing error: " + message,
 
                 line: ex.lineNumber,

--- a/lib/util/source-code.js
+++ b/lib/util/source-code.js
@@ -112,7 +112,7 @@ function SourceCode(text, ast) {
      * This is done to avoid each rule needing to do so separately.
      * @type string[]
      */
-    this.lines = text.split(/\r\n|\r|\n|\u2028|\u2029/g);
+    this.lines = SourceCode.splitLines(text);
 
     this.tokensAndComments = ast.tokens.concat(ast.comments).sort(function(left, right) {
         return left.range[0] - right.range[0];
@@ -132,6 +132,16 @@ function SourceCode(text, ast) {
     Object.freeze(this);
     Object.freeze(this.lines);
 }
+
+/**
+ * Split the source code into multiple lines based on the line delimiters
+ * @param {string} text Source code as a string
+ * @returns {string[]} Array of source code lines
+ * @public
+ */
+SourceCode.splitLines = function(text) {
+    return text.split(/\r\n|\r|\n|\u2028|\u2029/g);
+};
 
 SourceCode.prototype = {
     constructor: SourceCode,

--- a/tests/lib/eslint.js
+++ b/tests/lib/eslint.js
@@ -2160,6 +2160,22 @@ describe("eslint", function() {
             assert.equal(messages.length, 1);
             assert.equal(messages[0].severity, 2);
             assert.isNull(messages[0].ruleId);
+            assert.equal(messages[0].source, BROKEN_TEST_CODE);
+            assert.isTrue(messages[0].fatal);
+            assert.match(messages[0].message, /^Parsing error:/);
+        });
+
+        it("should report source code where the issue is present", function() {
+            var inValidCode = [
+                "var x = 20;",
+                "if (x ==4 {",
+                "    x++;",
+                "}"
+            ];
+            var messages = eslint.verify(inValidCode.join("\n"));
+            assert.equal(messages.length, 1);
+            assert.equal(messages[0].severity, 2);
+            assert.equal(messages[0].source, inValidCode[1]);
             assert.isTrue(messages[0].fatal);
             assert.match(messages[0].message, /^Parsing error:/);
         });
@@ -3202,6 +3218,7 @@ describe("eslint", function() {
                 var messages = eslint.verify(";", { parser: parser }, "filename");
                 assert.equal(messages.length, 1);
                 assert.equal(messages[0].severity, 2);
+                assert.isNull(messages[0].source);
                 assert.equal(messages[0].message, errorPrefix + require(parser).expectedError);
             });
 

